Build failure with `hashable-1.4.0.0`

Trying to build `parameterized-utils` with `hashable-1.4.0.0` fails thusly:

```
$ cabal build --allow-newer=hashable
Build profile: -w ghc-8.10.7 -O1
In order, the following will be built (use -v for more details):
 - parameterized-utils-2.1.4.0 (lib) (first run)
Preprocessing library for parameterized-utils-2.1.4.0..
Building library for parameterized-utils-2.1.4.0..
[ 3 of 35] Compiling Data.Parameterized.Classes

src/Data/Parameterized/Classes.hs:354:10: error:
    • Could not deduce (TestEquality f)
        arising from the superclasses of an instance declaration
      from the context: HashableF f
        bound by the instance declaration
        at src/Data/Parameterized/Classes.hs:354:10-46
      Possible fix:
        add (TestEquality f) to the context of the instance declaration
    • In the instance declaration for ‘Hashable (TypeAp f tp)’
    |
354 | instance HashableF f => Hashable (TypeAp f tp) where
    |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

This is because `hashable-1.4.0.0` makes `Eq` a superclass of `Hashable`. Because `TypeAp`'s `Eq` instance has a `TestEquality` constraint, the `Hashable` instance for `TypeAp` now also requires `TestEquality`. This isn't a problem currently, as `parameterized-utils` constraints `hashable < 1.4` in its `.cabal` file, but it will be an issue eventually.

Two possible fixes:

1. Add `TestEquality` constraints to all instances that need them. Something like this:

   <details>
   
   ```diff
   diff --git a/parameterized-utils.cabal b/parameterized-utils.cabal
   index 6df2300..aa86ded 100644
   --- a/parameterized-utils.cabal
   +++ b/parameterized-utils.cabal
   @@ -56,8 +56,8 @@ library
                   , containers
                   , deepseq
                   , ghc-prim
   -               , hashable       >=1.2  && <1.4
   -               , hashtables     ==1.2.*
   +               , hashable       >=1.2  && <1.5
   +               , hashtables     >=1.2  && <1.4
                   , indexed-traversable
                   , lens           >=4.16 && <5.1
                   , mtl
   @@ -142,4 +142,4 @@ test-suite parameterizedTests
    
      if impl(ghc >= 8.6)
        build-depends:
   -      hedgehog-classes
   \ No newline at end of file
   +      hedgehog-classes
   diff --git a/src/Data/Parameterized/Classes.hs b/src/Data/Parameterized/Classes.hs
   index b90754d..df1072a 100644
   --- a/src/Data/Parameterized/Classes.hs
   +++ b/src/Data/Parameterized/Classes.hs
   @@ -351,7 +351,7 @@ instance OrdF f => Ord (TypeAp f tp) where
    instance ShowF f => Show (TypeAp f tp) where
      showsPrec p (TypeAp x) = showsPrecF p x
    
   -instance HashableF f => Hashable (TypeAp f tp) where
   +instance (HashableF f, TestEquality f) => Hashable (TypeAp f tp) where
      hashWithSalt s (TypeAp x) = hashWithSaltF s x
    
    ------------------------------------------------------------------------
   diff --git a/src/Data/Parameterized/Context/Safe.hs b/src/Data/Parameterized/Context/Safe.hs
   index 96da3c0..5bdd196 100644
   --- a/src/Data/Parameterized/Context/Safe.hs
   +++ b/src/Data/Parameterized/Context/Safe.hs
   @@ -595,10 +595,10 @@ instance Hashable (Index ctx tp) where
    instance HashableF (Index ctx) where
      hashWithSaltF s i = hashWithSalt s (indexVal i)
    
   -instance HashableF f => HashableF (Assignment f) where
   +instance (HashableF f, TestEquality f) => HashableF (Assignment f) where
      hashWithSaltF = hashWithSalt
    
   -instance HashableF f => Hashable (Assignment f ctx) where
   +instance (HashableF f, TestEquality f) => Hashable (Assignment f ctx) where
      hashWithSalt s AssignmentEmpty = s
      hashWithSalt s (AssignmentExtend asgn x) = (s `hashWithSalt` asgn) `hashWithSaltF` x
    
   diff --git a/src/Data/Parameterized/Context/Unsafe.hs b/src/Data/Parameterized/Context/Unsafe.hs
   index 43e9def..e3f5845 100644
   --- a/src/Data/Parameterized/Context/Unsafe.hs
   +++ b/src/Data/Parameterized/Context/Unsafe.hs
   @@ -850,10 +850,10 @@ instance HashableF (Index ctx) where
    instance Hashable (Index ctx tp) where
      hashWithSalt = hashWithSaltF
    
   -instance HashableF f => Hashable (Assignment f ctx) where
   +instance (HashableF f, TestEquality f) => Hashable (Assignment f ctx) where
      hashWithSalt s (Assignment a) = hashWithSaltF s a
    
   -instance HashableF f => HashableF (Assignment f) where
   +instance (HashableF f, TestEquality f) => HashableF (Assignment f) where
      hashWithSaltF = hashWithSalt
    
    instance ShowF f => Show (Assignment f ctx) where
   diff --git a/src/Data/Parameterized/Some.hs b/src/Data/Parameterized/Some.hs
   index 3df9359..75e9c55 100644
   --- a/src/Data/Parameterized/Some.hs
   +++ b/src/Data/Parameterized/Some.hs
   @@ -33,7 +33,7 @@ instance TestEquality f => Eq (Some f) where
    instance OrdF f => Ord (Some f) where
      compare (Some x) (Some y) = toOrdering (compareF x y)
    
   -instance HashableF f => Hashable (Some f) where
   +instance (HashableF f, TestEquality f) => Hashable (Some f) where
      hashWithSalt s (Some x) = hashWithSaltF s x
      hash (Some x) = hashF x
   
   ```
   </details>
2. In the spirit of `hashable-1.4.0.0`, we could make `TestEquality` a superclass of `HashableF`. This comes with its own set of challenges, however, as there are data types that have `HashableF` instances but not `TestEquality` instances. For instance, naïvely adding a `TestEquality` instance to `HashableF` will result in a different build error in `parameterized-utils`:

   ```
   [ 3 of 35] Compiling Data.Parameterized.Classes
   
   src/Data/Parameterized/Classes.hs:331:10: error:
       • Could not deduce (TestEquality (Const a))
           arising from the superclasses of an instance declaration
         from the context: Hashable a
           bound by the instance declaration
           at src/Data/Parameterized/Classes.hs:331:10-42
       • In the instance declaration for ‘HashableF (Const a)’
       |
   331 | instance Hashable a => HashableF (Const a) where
       |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   ```
   
   It's not entirely clear to me how you'd even define `TestEquality` for `Const`.

--- a/Changelog.md
+++ b/Changelog.md
@@ -1,5 +1,13 @@
 # Changelog for the `parameterized-utils` package
 
+## next -- *TBA*
+
+  * Allow building with `hashable-1.4.*`. Because `hashable-1.4.0.0` adds an
+    `Eq` superclass to `Hashable`, some instances of `Hashable` in
+    `parameterized-utils` now require additional `TestEquality` constraints, as
+    the corresponding `Eq` instances for these data types also require
+    `TestEquality` constraints.
+
 ## 2.1.5.0 -- *2022 Mar 08*
 
   * Add support for GHC 9.2.  Drop support for GHC 8.4 (or earlier).

--- a/parameterized-utils.cabal
+++ b/parameterized-utils.cabal
@@ -56,8 +56,8 @@ library
                , containers
                , deepseq
                , ghc-prim
-               , hashable       >=1.2  && <1.4
-               , hashtables     ==1.2.*
+               , hashable       >=1.2  && <1.5
+               , hashtables     >=1.2  && <1.4
                , indexed-traversable
                , lens           >=4.16 && <5.2
                , mtl

--- a/src/Data/Parameterized/Classes.hs
+++ b/src/Data/Parameterized/Classes.hs
@@ -351,7 +351,7 @@ instance OrdF f => Ord (TypeAp f tp) where
 instance ShowF f => Show (TypeAp f tp) where
   showsPrec p (TypeAp x) = showsPrecF p x
 
-instance HashableF f => Hashable (TypeAp f tp) where
+instance (HashableF f, TestEquality f) => Hashable (TypeAp f tp) where
   hashWithSalt s (TypeAp x) = hashWithSaltF s x
 
 ------------------------------------------------------------------------

--- a/src/Data/Parameterized/Context/Safe.hs
+++ b/src/Data/Parameterized/Context/Safe.hs
@@ -595,10 +595,10 @@ instance Hashable (Index ctx tp) where
 instance HashableF (Index ctx) where
   hashWithSaltF s i = hashWithSalt s (indexVal i)
 
-instance HashableF f => HashableF (Assignment f) where
+instance (HashableF f, TestEquality f) => HashableF (Assignment f) where
   hashWithSaltF = hashWithSalt
 
-instance HashableF f => Hashable (Assignment f ctx) where
+instance (HashableF f, TestEquality f) => Hashable (Assignment f ctx) where
   hashWithSalt s AssignmentEmpty = s
   hashWithSalt s (AssignmentExtend asgn x) = (s `hashWithSalt` asgn) `hashWithSaltF` x
 

--- a/src/Data/Parameterized/Context/Unsafe.hs
+++ b/src/Data/Parameterized/Context/Unsafe.hs
@@ -851,10 +851,10 @@ instance HashableF (Index ctx) where
 instance Hashable (Index ctx tp) where
   hashWithSalt = hashWithSaltF
 
-instance HashableF f => Hashable (Assignment f ctx) where
+instance (HashableF f, TestEquality f) => Hashable (Assignment f ctx) where
   hashWithSalt s (Assignment a) = hashWithSaltF s a
 
-instance HashableF f => HashableF (Assignment f) where
+instance (HashableF f, TestEquality f) => HashableF (Assignment f) where
   hashWithSaltF = hashWithSalt
 
 instance ShowF f => Show (Assignment f ctx) where

--- a/src/Data/Parameterized/Some.hs
+++ b/src/Data/Parameterized/Some.hs
@@ -33,7 +33,7 @@ instance TestEquality f => Eq (Some f) where
 instance OrdF f => Ord (Some f) where
   compare (Some x) (Some y) = toOrdering (compareF x y)
 
-instance HashableF f => Hashable (Some f) where
+instance (HashableF f, TestEquality f) => Hashable (Some f) where
   hashWithSalt s (Some x) = hashWithSaltF s x
   hash (Some x) = hashF x
 

